lib/trycourier.rb

```diff
@@ -18,49 +29,75 @@ def initialize(code, message_id)
   end
 
   class Client
-    def initialize(auth_token = nil)
-      @auth_token = auth_token || ENV['COURIER_AUTH_TOKEN']
-      @uri = URI.parse('https://api.trycourier.app/send')
-      if @auth_token == nil or @auth_token == ""
-        raise InputError, "Client requires an auth_token be supplied."
+    def initialize(auth_token: nil, base_url: nil, username: nil, password: nil)
+      @base_url = if base_url
+        base_url
+      elsif ENV["COURIER_BASE_URL"]
+        ENV["COURIER_BASE_URL"]
+      else
+        "https://api.courier.com"
       end
+
+      @session = Courier::CourierAPISession.new
+
+      if auth_token
+        @session.init_token_auth(auth_token)
+      elsif ENV["COURIER_AUTH_TOKEN"]
+        @session.init_token_auth(ENV["COURIER_AUTH_TOKEN"])
+      elsif username && password
+        @session.init_basic_auth(username, password)
+      elsif ENV["COURIER_AUTH_USERNAME"] && ENV["COURIER_AUTH_PASSWORD"]
+        @session.init_basic_auth(ENV["COURIER_AUTH_USERNAME"], ENV["COURIER_AUTH_PASSWORD"])
+      end
+
+      @messages = Courier::Messages.new(@base_url, @session)
```

Comment:
Why is the `base_url` not part of the `Courier::CourierAPISession`? It feels like it should be included like the auth is and then we don't have to pass it around extra.